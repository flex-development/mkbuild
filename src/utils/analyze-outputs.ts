/**
 * @file Utilities - analyzeOutputs
 * @module mkbuild/utils/analyzeOutputs
 */

import type { OutputMetadata } from '#src/types'
import * as color from 'colorette'
import pb from 'pretty-bytes'
import { template } from 'radash'

/**
 * Generates a build analysis for the given `outputs`.
 *
 * @param {string} outdir - Output directory name
 * @param {Record<string, Pick<OutputMetadata, 'bytes'>>} [outputs={}] - Build
 * output map from metafile
 * @return {{ analysis: string; bytes: number }} Build output analysis and size
 */
const analyzeOutputs = (
  outdir: string,
  outputs: Record<string, Pick<OutputMetadata, 'bytes'>> = {}
): { analysis: string; bytes: number } => {
  /**
   * Indentation.
   *
   * @const {string} indent
   */
  const indent: string = ' '.repeat(2)

  /**
   * Output strings.
   *
   * @const {string[]} strings
   */
  const strings: string[] = []

  /**
   * Total output size.
   *
   * @const {number} size
   */
  const size: number = Object.entries(outputs).reduce((acc, output) => {
    const [outfile, metadata] = output
    strings.push(color.gray(`${indent}└─ ${outfile} (${pb(metadata.bytes)})`))
    return acc + metadata.bytes
  }, 0)

  return {
    analysis: template(`${indent}{{0}} (total size: {{1}})\n{{2}}`, {
      0: color.bold(outdir),
      1: color.cyan(pb(size)),
      2: strings.join('\n')
    }),
    bytes: size
  }
}

export default analyzeOutputs
