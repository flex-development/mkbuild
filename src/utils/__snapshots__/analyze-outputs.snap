// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/analyzeOutputs > should return build analysis object 1`] = `
"  [1mdist[22m (total size: [36m4.48 kB[39m)
[90m  â””â”€ dist/index.d.mts (86 B)[39m
[90m  â””â”€ dist/index.mjs (30 B)[39m
[90m  â””â”€ dist/interfaces/entry.d.mts (342 B)[39m
[90m  â””â”€ dist/interfaces/entry.mjs (0 B)[39m
[90m  â””â”€ dist/interfaces/index.d.mts (342 B)[39m
[90m  â””â”€ dist/interfaces/index.mjs (0 B)[39m
[90m  â””â”€ dist/interfaces/result.d.mts (1.21 kB)[39m
[90m  â””â”€ dist/interfaces/result.mjs (0 B)[39m
[90m  â””â”€ dist/interfaces/source-file.d.mts (583 B)[39m
[90m  â””â”€ dist/interfaces/source-file.mjs (0 B)[39m
[90m  â””â”€ dist/interfaces/statement.d.mts (686 B)[39m
[90m  â””â”€ dist/interfaces/statement.mjs (0 B)[39m
[90m  â””â”€ dist/utils/analyze-outputs.d.mts (626 B)[39m
[90m  â””â”€ dist/utils/analyze-outputs.mjs (575 B)[39m"
`;
